src/EntityRegistrantDetection/CallbackWrapper.php

```diff
@@ -0,0 +1,347 @@
+	 * @var array
+	 */
+	protected $source;
+
+	/**
+	 * List of registered entities by callback.
+	 *
+	 * @var array [
+	 *     @type array $post_type List of registered post types slug.
+	 *     @type array $taxonomy  List of registered taxonomy slug.
+	 *     @type array $block     List of registered block type.
+	 *     @type array $shortcode List of registered shortcode.
+	 * ]
+	 */
+	protected $registered_entities;
+
+	/**
+	 * AMP_Validation_Callback_Wrapper constructor.
+	 *
+	 * @param array $callback [
+	 *     The callback data.
+	 *     @type callable $function
+	 *     @type int      $accepted_args
+	 *     @type int      $priority
+	 *     @type string   $hook
+	 * ]
+	 */
+	public function __construct( $callback ) {
+
+		$this->callback            = $callback;
+		$this->registered_entities = [
+			'post_type' => [],
+			'taxonomy'  => [],
+			'block'     => [],
+			'shortcode' => [],
+		];
+	}
+
+	/**
+	 * Get callback function.
+	 *
+	 * @return callable
+	 */
+	public function get_callback_function() {
+
+		return $this->callback['function'];
+	}
+
+	/**
+	 * Invoke wrapped callback.
+	 *
+	 * @param array ...$args Args.
+	 *
+	 * @return mixed Response.
+	 */
+	public function __invoke( ...$args ) {
+
+		$this->prepare();
+
+		$result = call_user_func_array(
+			$this->get_callback_function(),
+			array_slice( $args, 0, (int) $this->callback['accepted_args'] )
+		);
+
+		$this->finalize();
+
+		return $result;
+	}
+
+	/**
+	 * Set the source of a callback function.
+	 *
+	 * @return void
+	 */
+	protected function set_source() {
+
+		$callback_reflection = Services::get( 'dev_tools.callback_reflection' );
+		$this->source        = $callback_reflection->get_source( $this->get_callback_function() );
+
+		unset( $this->source['reflection'] );
+
+		if ( $this->callback['hook'] ) {
+			$this->source['hook'] = $this->callback['hook'];
+		}
+
+		if ( $this->callback['priority'] ) {
+			$this->source['priority'] = $this->callback['priority'];
+		}
+	}
+
+	/**
+	 * Collect the currently registered entities before executing the original function.
+	 *
+	 * @return void
+	 */
+	protected function prepare() {
+
+		$this->registered_entities = $this->get_registered_entities();
+	}
+
+	/**
+	 * Find the difference of registered entities between before and after executing function.
+	 *
+	 * @return void
+	 */
+	protected function finalize() {
+
+		$final_registered_entities = $this->get_registered_entities();
+
+		foreach ( $this->registered_entities as $entity_type => $entities ) {
+
+			$different = array_diff(
+				$final_registered_entities[ $entity_type ],
+				$this->registered_entities[ $entity_type ],
+				self::DEFAULT_ENTITIES[ $entity_type ]
+			);
+			$different = array_values( $different );
+
+			if ( $different && empty( $this->source ) ) {
+				$this->set_source();
+			}
+
+			$this->registered_entities[ $entity_type ] = $different;
+
+			EntityRegistrantDetectionManager::add_source( $entity_type, $different, $this->source );
```

Comment:
We should avoid static method usage.

Reply:
I realize that complicates the architecture, but maybe not so much. If `EntityRegistrantDetectionManager` is simply added among the services returned by `\AmpProject\AmpWP\AmpWpPlugin::get_shared_instances()`, then only one copy of `EntityRegistrantDetectionManager` will be instantiated and this will then have the same effect.